src/gf.c

```diff
@@ -134,6 +142,13 @@ static inline int cache_match(jl_value_t **args, size_t n, jl_tupletype_t *sig,
                 // to match exactly either. this is cached as Type{T}.
                 // analogous to the situation with tuples.
             }
+            else if (jl_is_typector(a)) {
+                // in the case where a is exactly a
+                // TypeConstructor, tp0 must be exactly
+                // Type{TypeConstructor}
+                if (tp0 != (jl_value_t*)jl_typector_type)
```

Comment:
This doesn't seem right to me. `isa(a, Type{TypeConstructor})` is only true if `a === TypeConstructor`, not if `isa(a, TypeConstructor)`. In fact, if I remove this and the corresponding bit of `cache_match_by_type`, the tests still pass.